### PR TITLE
Admin API decorators

### PR DESCRIPTION
**Pull request by sravfeyn**
Started working on this because of http://manage.dimagi.com/default.asp?231858#1182164 , but uncovered more bugs on the way and some unclean code. This PR fixes some bugs and refactors bunch of code. Best viewed commit by commit
@biyeun

### CODE REVIEW

**Comment by sravfeyn** [comment]
Biyeun is offline, @czue @millerdev Pinging you for review since both of you have context on this

File: corehq/apps/api/tests.py [hunk 1]
```diff
@@ -1608,6 +1615,36 @@ def test_basic(self):
         self.assert_accessible_via_sessions(self.list_endpoint)
 
 
+class AdminWebResourceTest(APIResourceTest):
+    resource = v0_5.AdminWebUserResource
+    api_name = 'global'
+
+    def tearDown(self):
+        self.community_domain.delete()
+        self.new_user.delete()
```
Comment (czue) [inline corehq/apps/api/tests.py]: I think if you create these objects in a test it's safer to use `addCleanup` in that same test

File: corehq/apps/domain/decorators.py [hunk 2]
```diff
@@ -147,6 +147,36 @@ def wrapper(request, *args, **kwargs):
...
+                domain_name, domain = load_domain(request, args[0])
+                if domain:
+                    raise Exception(
+                        "This decorator doesn't do domain-specific validation. Please use decorators "
+                        "based on _login_or_challenge for domain-specific auth and validation"
+                    )
+
+            @check_lockout
+            @challenge_fn
+            @require_superuser_or_developer
```
Comment (czue) [inline corehq/apps/domain/decorators.py]: is `or_developer` right? i think it's consistent with the prior code, but if that's the intention it might be better to include it in the name of the function

**Comment by czue** [comment]
looks good. I don't feel familiar enough with the tastypie objects to understand any edge-cases or other possible issues that could come from this. how confident are you in testing and/or test coverage for this?

**Comment by millerdev** [comment]
agree with @czue's comments. otherwise looks ok to me

File: corehq/apps/domain/decorators.py (hunk 2 above)
Comment (sravfeyn) [inline corehq/apps/domain/decorators.py]: This is what the old `AdminAuthentication` looked for (including `or_developer`)

File: corehq/apps/api/tests.py (hunk 1 above)
Comment (sravfeyn) [inline corehq/apps/api/tests.py]: Nice! (added)

**Comment by sravfeyn** [comment]
I am pretty confident about the test coverage as I have added resource testing with session, with API key, with normal user and with admin user for all admin APIs. There is one resource that is failing (`DefaultProductPlanResource`) but for a different reason. It looks like an unrelated [bug on prod](http://manage.dimagi.com/default.asp?232692_9ghna3jvshp5prb0)

**Comment by sravfeyn** [comment]
I checked on proxy log to see who uses these APIs. Almost all of the requests are from [this user](https://www.commcarehq.org/hq/admin/web_user_lookup/?q=niketsoral%40gmail.com). He is not an admin/superuser, so he shouldn't be able to access these on first sight. But he is able to, because he is part of domain called 'dimagi' and the old [hackish implementation](https://github.com/dimagi/commcare-hq/pull/12470/files#diff-abc8144aa45ba1dba5b00e9b1169b394L168) of admin api auth allows users of 'dimagi' project space to access admin APIs (which this PR removes). All of these requests from this user will fail after the PR, because he is not an admin.

I am not sure why this user is in that domain but he was able to access bunch of admin api URLs

**Comment by czue** [comment]
I think niket is a salesforce developer and that account might be used for the SF integration. might want to check with @fyniac in case it's important that he can still access those.

**Comment by fyniac** [comment]
Hey, we do still need access to the admin APIs for the HQ-SF work.  Likely changing access to Lukasz though, but I think he has similar permissions

**Comment by sravfeyn** [comment]
Cool. Even if we have to continue to give access, we should probably change the way we grant admin API access permissions.

Currently the hackish way is to add everyone in domain 'dimagi' to allow admin API access. Should we continue to do it this way or add a new 'Admin API access role' that users can be assigned to by superusers? @czue

**Comment by czue** [comment]
agreed. I was under the impression that was what the `is_developer` check was for. can we just use that?

**Comment by sravfeyn** [comment]
Oh yeah, that's right. It looks like niket is on that [toggle already](https://www.commcarehq.org/hq/flags/edit/is_developer/). So, then this shouldn't break anything for him.

**Comment by czue** [comment]
cool, I think this is g2g on :white_check_mark:

**Comment by czue** [comment]
build fail:
```
======================================================================
ERROR: test_basic (corehq.apps.api.tests.AdminAPITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 1648, in test_basic
    self.assert_admin_resource(url)
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 1632, in assert_admin_resource
    response = self._assert_auth_get_resource(url, username='admin', password='pass')
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 186, in _assert_auth_get_resource
    response = self.client.get(api_url)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 470, in get
    **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 286, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 358, in generic
    return self.request(**r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 440, in request
    six.reraise(*exc_info)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 219, in wrapper
    response = callback(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 450, in dispatch_list
    return self.dispatch('list', request, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/api/domain_metadata.py", line 41, in dispatch
    return Resource.dispatch(self, request_type, request, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 482, in dispatch
    response = method(request, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 1335, in get_list
    to_be_serialized = paginator.page()
  File "/vendor/lib/python2.7/site-packages/tastypie/paginator.py", line 194, in page
    count = self.get_count()
  File "/vendor/lib/python2.7/site-packages/tastypie/paginator.py", line 126, in get_count
    return self.objects.count()
  File "/mnt/commcare-hq-ro/corehq/apps/api/domain_metadata.py", line 26, in count
    return self.es_query.size(0).run().total
  File "/mnt/commcare-hq-ro/corehq/apps/es/es_query.py", line 186, in run
    raw = run_query(self.index, self.raw_query)
  File "/mnt/commcare-hq-ro/corehq/elastic.py", line 145, in run_query
    return get_es_new().search(es_meta.index, es_meta.type, body=q)
  File "/vendor/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 69, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/vendor/lib/python2.7/site-packages/elasticsearch/client/__init__.py", line 531, in search
    doc_type, '_search'), params=params, body=body)
  File "/vendor/lib/python2.7/site-packages/elasticsearch/transport.py", line 307, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, ignore=ignore, timeout=timeout)
  File "/vendor/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 93, in perform_request
    self._raise_error(response.status, raw_data)
  File "/vendor/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 105, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
NotFoundError: TransportError(404, u'IndexMissingException[[test_hqdomains_20160627_1351] missing]')
-------------------- >> begin captured logging << --------------------
```

**Comment by biyeun** [comment]
@sravfeyn conflicts

**Comment by sravfeyn** [comment]
closing in favor of https://github.com/dimagi/commcare-hq/pull/13648
